src/azure-cli-core/azure/cli/core/util.py

```diff
@@ -74,78 +73,165 @@ def handle_exception(ex):  # pylint: disable=too-many-return-statements, too-man
+            if is_azure_connection_error(error_msg):
+                az_error = azclierror.AzureConnectionError(error_msg)
+            else:
+                az_error = azclierror.ClientRequestError(error_msg)
 
-        # TODO: Fine-grained analysis
-        elif isinstance(ex, (ClientRequestError, SSLError)):
-            az_error = AzCLIError(AzCLIErrorType.ClientError, error_msg)
-            if 'SSLError' in error_msg:
-                az_error.set_recommendation(SSLERROR_TEMPLATE)
+        elif isinstance(ex, ClientRequestError):
+            if is_azure_connection_error(error_msg):
+                az_error = azclierror.AzureConnectionError(error_msg)
+            else:
+                az_error = azclierror.ClientRequestError(error_msg)
 
-        # TODO: Fine-grained analysis
         elif isinstance(ex, HttpOperationError):
-            try:
-                response = json.loads(ex.response.text)
-                if isinstance(response, str):
-                    error = response
-                else:
-                    error = response['error']
-
-                # ARM should use ODATA v4. So should try this first.
-                # http://docs.oasis-open.org/odata/odata-json-format/v4.0/os/odata-json-format-v4.0-os.html#_Toc372793091
-                if isinstance(error, dict):
-                    code = "{} - ".format(error.get('code', 'Unknown Code'))
-                    message = error.get('message', ex)
-                    error_msg = "code: {}, {}".format(code, message)
-                else:
-                    error_msg = error
-
-            except (ValueError, KeyError):
-                pass
-
-            az_error = AzCLIError(AzCLIErrorType.ServiceError, error_msg)
+            if extract_http_operation_error_message(ex):
+                error_msg = extract_http_operation_error_message(ex)
+            # by HttpOperationError definition, the message should be with the
+            # format: '({code}) {message}' if there is a status code
+            status_code = ex.__str__().split(')')[0].split('(')[-1]
+            AzCLIErrorType = get_error_type_by_status_code(status_code)
+            az_error = AzCLIErrorType(error_msg)
 
         elif isinstance(ex, KeyboardInterrupt):
             error_msg = 'Keyboard interrupt is captured.'
-            az_error = AzCLIError(AzCLIErrorType.ManualInterrupt, error_msg)
+            az_error = azclierror.ManualInterrupt(error_msg)
 
         else:
             error_msg = "The command failed with an unexpected error. Here is the traceback:"
-            az_error = AzCLIError(AzCLIErrorType.UnexpectedError, error_msg)
-            az_error.set_raw_exception(ex)
+            az_error = azclierror.CLIInternalError(error_msg)
+            az_error.set_exception_trace(ex)
             az_error.set_recommendation("To open an issue, please run: 'az feedback'")
 
+        if isinstance(az_error, azclierror.ResourceNotFoundError):
+            exit_code = 3
```

Comment:
Should exit_code be 3 for common resource-not-found cases? Or only for show command?

Reply:
The definition is ambiguous.
https://github.com/Azure/azure-cli#exit-codes
I think 3 is OK.

Reply:
what's the original exit code?

Reply:
3 is good, it's not limited to show command.

Reply:
The original one is magic, it even returns 'CloudError' as an exit_code and even check it in testing file [here](https://github.com/Azure/azure-cli/pull/15319/files#diff-96f657597aa99f9b338dc445b0d45590L429-L439). Could not understand :(